Migrate "Customers > Outstanding"

ks129 opened the pull request:
<!-----------------------------------------------------------------------------
Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows below with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop.com/1.7/contribute/contribution-guidelines/#pull-requests
------------------------------------------------------------------------------>

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Migrate old legacy outstanding controller (and other components) to modern controller.
| Type?         | refacto
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Fixes #17156.
| How to test?  | After enabling B2B, "Customers > Outstanding" page should use modern controller. ![Screenshot 2020-07-23 at 10 19 16](https://user-images.githubusercontent.com/45097959/88260943-0a846700-ccce-11ea-9d05-ccc040467b5d.png)


<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/20288)
<!-- Reviewable:end -->

ks129 committed:
Update BadgeColumn to accept custom colors

ks129 committed:
Add outstanding query builder for grid

ks129 committed:
Add outstanding grid data factory

ks129 committed:
Add outstanding grid definition factory

ks129 committed:
Add outstanding grid filters

ks129 committed:
Add new modern outstanding controller

ks129 committed:
Add Twig index file for outstanding page

ks129 committed:
Add JS index file for outstanding page

ks129 committed:
Add outstanding grid information to YML files

ks129 committed:
Add outstanding routes

ks129 committed:
Add outstanding JS file to common.js

ks129 committed:
Remove old legacy outstanding controller + PDF icon template

ks129 committed:
Remove legacy outstanding controller from unit test

PululuK commented on src/Core/Grid/Column/Type/Common/BadgeColumn.php:
```diff
@@ -54,13 +54,14 @@ protected function configureOptions(OptionsResolver $resolver)
             ])
             ->setDefaults([
-                'badge_type' => 'success',
+                'badge_type' => '',
```
Hi @ks129 

I totally agree with you, but it is outside the scope of your PR and it can be considered a BC.
Instead, you can keep the current behavior and set these options here:
`src/Core/Grid/Definition/Factory/OutstandingGridDefinitionFactory.php
`
And then if you want you can create another PR.

Progi1984 commented on admin-dev/themes/new-theme/.webpack/common.js:
```diff
@@ -85,6 +85,7 @@ module.exports = {
     order_message: './js/pages/order_message',
     order_preferences: './js/pages/order-preferences',
     order_view: './js/pages/order/view.js',
+    outstanding: './js/pages/outstanding',
```
```suggestion
    customer_outstanding: './js/pages/outstanding',
```

PululuK commented on src/Core/Grid/Data/Factory/OutstandingGridDataFactory.php:
```diff
@@ -0,0 +1,101 @@
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
```
Don't forget declare(strict_types=1); on new files.
See https://github.com/PrestaShop/docs/pull/518/files

Progi1984 commented on src/Core/Grid/Data/Factory/OutstandingGridDataFactory.php:
```diff
@@ -0,0 +1,101 @@
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+namespace PrestaShop\PrestaShop\Core\Grid\Data\Factory;
```
```suggestion
 */
 
 declare(strict_types=1);

namespace PrestaShop\PrestaShop\Core\Grid\Data\Factory;
```

Progi1984 commented on src/Core/Grid/Data/Factory/OutstandingGridDataFactory.php:
```diff
@@ -0,0 +1,101 @@
+    public function __construct(
+        GridDataFactoryInterface $outstandingDataFactory,
+        RepositoryInterface $repositoryLocale,
+        $contextLocale
```
```suggestion
        string $contextLocale
```

Progi1984 commented on src/Core/Grid/Data/Factory/OutstandingGridDataFactory.php:
```diff
@@ -0,0 +1,101 @@
+
+        foreach ($records as &$record) {
+            $customer = new Customer((int) $record['id_customer']);
+            if (Validate::isLoadedObject($customer)) {
+                $record['outstanding'] = $locale->formatPrice($customer->getOutstanding(), $record['iso_code']);
+            } else {
+                $record['outstanding'] = $locale->formatPrice(0.00, $record['iso_code']);
+            }
```
```suggestion
            $record['outstanding'] = $locale->formatPrice(
                Validate::isLoadedObject($customer) ? $customer->getOutstanding() : 0.00,
                $record['iso_code']
            );
```

PululuK commented:
Great work  👌🏽
Thanks @ks129

Progi1984 commented on src/Core/Grid/Definition/Factory/OutstandingGridDefinitionFactory.php:
```diff
@@ -0,0 +1,240 @@
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+namespace PrestaShop\PrestaShop\Core\Grid\Definition\Factory;
```
```suggestion
 */

declare(strict_types=1);

namespace PrestaShop\PrestaShop\Core\Grid\Definition\Factory;
```

ks129 committed:
Add strict typing to outstanding files

ks129 committed:
Rename outstanding web pack common.js definition

ks129 committed:
Remove badge column BC change

ks129 committed:
Simplify outstanding formatting

ks129 committed:
Only allow downloading invoice in outstanding when this is enabled

ks129 commented:
Why PHPStan is failing? I don't understand these errors.

ks129 committed:
Move outstanding files declare statement to very top of file

Progi1984 commented on src/PrestaShopBundle/Resources/views/Admin/Common/Grid/Columns/Content/badge.html.twig:
```diff
@@ -24,10 +24,16 @@
     {{ record[column.options.field] }}
   </span>
+{% elseif record[column.options.field] is not empty and column.options.color_field in record|keys and record[column.options.color_field] is not empty %}
+  <span class="badge rounded" style="background-color: {{ record[column.options.color_field] }}">
```
```suggestion
  {% set textColor = is_color_bright(record[column.options.color_field]) ? '#383838' : 'white' %}
  <span class="badge rounded" style="background-color: {{ record[column.options.color_field] }}; color: {{ textColor ]]">
```

Progi1984 reviewed (commented):
A small change

Progi1984 commented:
@ks129 Very good job. It's a great contribution to the project.

ks129 committed:
Make badge column handling text colour

Co-authored-by: Progi1984 <progi1984@gmail.com>

Progi1984 commented on src/PrestaShopBundle/Resources/views/Admin/Common/Grid/Columns/Content/badge.html.twig:
```diff
@@ -24,12 +24,18 @@
   </span>
+{% elseif record[column.options.field] is not empty and column.options.color_field in record|keys and record[column.options.color_field] is not empty %}
+  {% set textColor = is_color_bright(record[column.options.color_field]) ? '#383838' : 'white' %}
+  <span class="badge rounded" style="background-color: {{ record[column.options.color_field] }}; color: {{ textColor ]]">
```
```suggestion
  <span class="badge rounded" style="background-color: {{ record[column.options.color_field] }}; color: {{ textColor }}">
```

Progi1984 committed:
Fixed Travis CI

Junebyun commented on src/Core/Grid/Definition/Factory/OutstandingGridDefinitionFactory.php:
```diff
@@ -0,0 +1,263 @@
+            )
+            ->add(
+                (new DataColumn('outstanding_allow_amount'))
+                    ->setName($this->trans('Outstanding Allowance', [], 'Admin.Global'))
```
```suggestion
                    ->setName($this->trans('Outstanding allowance', [], 'Admin.Global'))
```

Junebyun replied:
```suggestion
                    ->setName($this->trans('Outstanding allowance', [], 'Admin.Global'))
```

Junebyun commented on src/PrestaShopBundle/Controller/Admin/Sell/Customer/OutstandingController.php:
```diff
@@ -0,0 +1,87 @@
+            [
+                'layoutHeaderToolbarBtn' => [],
+                'outstandingGrid' => $this->presentGrid($grid),
+                'layoutTitle' => $this->trans('Outstanding', 'Admin.Action'),
```
```suggestion
                'layoutTitle' => $this->trans('Outstanding', 'Admin.Navigation.Menu'),
```

Junebyun commented on src/Core/Grid/Definition/Factory/OutstandingGridDefinitionFactory.php:
```diff
@@ -0,0 +1,263 @@
+            )
+            ->add(
+                (new DataColumn('outstanding'))
+                    ->setName($this->trans('Current Outstanding', [], 'Admin.Global'))
```
```suggestion
                    ->setName($this->trans('Current outstanding', [], 'Admin.Global'))
```

ks129 committed:
Fix outstanding page layout title translation domain

Co-authored-by: Junebyun <64894223+Junebyun@users.noreply.github.com>

ks129 committed:
Make outstanding grid column key second words lowercase

Co-authored-by: Junebyun <64894223+Junebyun@users.noreply.github.com>

LouiseBonnard commented on src/Core/Grid/Definition/Factory/OutstandingGridDefinitionFactory.php:
```diff
@@ -0,0 +1,263 @@
+        return (new ColumnCollection())
+            ->add(
+                (new DataColumn('id_invoice'))
+                    ->setName($this->trans('Invoice', [], 'Admin.Global'))
```
Hi there, why not calling this column _ID_?

ks129 replied:
I named this `Invoice` because old controller used this instead `ID`.

LouiseBonnard replied:
In my opinion, it makes more sense to have an ID column for several reasons:

1. the search bar states 'Search ID'
2. it displays ID numbers and not invoice numbers

ks129 replied:
Okay, changed this to ID.

LouiseBonnard replied:
Nice, thanks!

LouiseBonnard commented on src/Core/Grid/Definition/Factory/OutstandingGridDefinitionFactory.php:
```diff
@@ -0,0 +1,263 @@
+            )
+            ->add(
+                (new BadgeColumn('risk'))
+                    ->setName($this->trans('Risk', [], 'Admin.Global'))
```
```suggestion
                    ->setName($this->trans('Risk', [], 'Admin.Orderscustomers.Feature'))
```

LouiseBonnard commented on src/Core/Grid/Definition/Factory/OutstandingGridDefinitionFactory.php:
```diff
@@ -0,0 +1,263 @@
+            )
+            ->add(
+                (new DataColumn('outstanding_allow_amount'))
+                    ->setName($this->trans('Outstanding allowance', [], 'Admin.Global'))
```
```suggestion
                    ->setName($this->trans('Outstanding allowance', [], 'Admin.Orderscustomers.Feature'))
```

LouiseBonnard commented on src/Core/Grid/Definition/Factory/OutstandingGridDefinitionFactory.php:
```diff
@@ -0,0 +1,263 @@
+            )
+            ->add(
+                (new DataColumn('outstanding'))
+                    ->setName($this->trans('Current outstanding', [], 'Admin.Global'))
```
```suggestion
                    ->setName($this->trans('Current outstanding', [], 'Admin.Orderscustomers.Feature'))
```

ks129 committed:
Fix outstanding grid translation domains

Co-authored-by: LouiseBonnard <32565890+LouiseBonnard@users.noreply.github.com>

ks129 committed:
Rename outstanding grid invoice column to ID

ks129 commented:
Why CI is failing?

matks commented:
Closing and reopening to restart travis 🤔  but it does not seem to do the job

matks committed:
Remove useless phpDoc

matks commented:
I added a useless commit https://github.com/PrestaShop/PrestaShop/pull/20288/commits/cb9fe88a3f94f329fe6bc3da6bb3e2dcf7c581ae to restart CI

matks commented on src/Core/Grid/Definition/Factory/OutstandingGridDefinitionFactory.php:
```diff
@@ -0,0 +1,263 @@
+    /**
+     * @param ConfigurationInterface $configuration
+     */
+    public function __construct(ConfigurationInterface $configuration)
```
This constructor is not compatible with `AbstractGridDefinitionFactory ` constructor
```
    public function __construct(HookDispatcherInterface $hookDispatcher = null)
```

You will get a
```
Type error: Argument 1 passed to PrestaShop\PrestaShop\Core\Grid\Definition\Factory\OutstandingGridDefinitionFactory::__construct() must implement interface PrestaShop\PrestaShop\Core\ConfigurationInterface, instance of PrestaShop\PrestaShop\Core\Hook\HookDispatcher given
```

matks reviewed (changes requested):
Hi @ks129 you did a really good job !

The code is overall very good 👍  however when I tried the PR I got an issue related to the OutstandingGridDefinitionFactory constructor (see comment).

Please fix this tiny little issue and then PR can get approved and move to QA phase 😉

ks129 committed:
Fix outstanding grid definition factory construct

matks commented:
Hi @ks129 ! PR is good 👍  however there was a git conflict. In order to fix it, I rebased your PR on develop* so I had to `git push --force` your PR.

PR can now move to QA 😉 

*If you dont know what this means, see https://dev.to/mathieuks/what-it-means-to-rebase-a-pull-request-submitted-on-github-5717

PierreRambaud commented on src/Core/Grid/Definition/Factory/OutstandingGridDefinitionFactory.php:
```diff
@@ -0,0 +1,267 @@
+    protected function getFilters()
+    {
+        $risks = [];
+        foreach (Risk::getRisks(Context::getContext()->language->id) as $risk) {
```
At least the context must be in the constructor.
If possible, even `Risk::getRisks`

Progi1984 replied:
@ks129 Comment not applied.

PierreRambaud commented on tests-legacy/PrestaShopBundle/Controller/ControllerTest.php:
```diff
@@ -133,7 +133,6 @@ public function getControllersClasses()
             array('AdminShopController'),
             array('AdminCartRulesController'),
             array('AdminGroupsController'),
-            array('AdminOutstandingController'),
```
We should add it in SurvivalTest no? :)

khouloudbelguith commented:
Hi @ks129,

I found those issues testing this PR
1. The sort buttons do not work for all field
![image](https://user-images.githubusercontent.com/16067358/92132490-81ea0200-ee07-11ea-9f2f-12b9ebf1e190.png)
2. In legacy, there is no **sort** action and **filter** box for "Current outstanding", also there is no sort action for the "Risk"
**Before**
![image](https://user-images.githubusercontent.com/16067358/92132976-1f453600-ee08-11ea-9348-358e9a94aa19.png)
**After**
![image](https://user-images.githubusercontent.com/16067358/92133078-413eb880-ee08-11ea-901b-978bd94c8e2a.png)
![image](https://user-images.githubusercontent.com/16067358/92133136-5287c500-ee08-11ea-9e71-b4d8d5dc6164.png)
3. A difference when the Risk is not defined for a customer
**Before**
After

ks129 commented:
@khouloudbelguith 
- Filter reset and sorting issues is now fixed by adding outstanding JS to webpack config.
- Removed sort and filter from "Current outstanding" and removed sort from "Risk".
- Made that in grid, "Risk" is shown as `--` when this don't exist for customer, but I can't do anything with dropdown: This `-` was automatically added by Legacy controller what means that this don't use this filter for filtering (there was no way to filter for invoices what customer don't have risk defined), but for modern grid, this automatically add empty entry instead. 
- Added support for multistore.
- Fixed wrong route that gave delivery slip instead of invoice.
- This don't show error in console for me.

Progi1984 commented:
@ks129 We are waiting for an answer from the @PrestaShop/prestashop-product-team (https://github.com/PrestaShop/PrestaShop/pull/20288#issuecomment-686560282 : N°4)

LouiseBonnard commented:
Alright, @khouloudbelguith, sorry for the late reply. Actually, this page is displayed only when the multistore mode is enabled. :-)

ks129 commented:
@LouiseBonnard This page is shown when B2B is enabled, not multistore.

LouiseBonnard commented:
You're right, my bad!

matks reviewed (dismissed):
I checked the latest changes, the code is good 👍

khouloudbelguith commented:
Hi @ks129,

There are small differences between the legacy page and the new page
1. The text-align in the new page is not in the center
**Before**
![image](https://user-images.githubusercontent.com/16067358/93222165-f3fb0900-f76e-11ea-8005-6a7e28e5a918.png)
**After**
![image](https://user-images.githubusercontent.com/16067358/93222250-0a08c980-f76f-11ea-8254-1ba5490e7eb7.png)

2. When using a mobile
**Before:** all columns are displayed
![image](https://user-images.githubusercontent.com/16067358/93222837-acc14800-f76f-11ea-96b4-6084ba7d1574.png)

**After:** Risk, Outstanding allowance, Current outstanding, Invoice and Actions columns are not displayed
![image](https://user-images.githubusercontent.com/16067358/93222889-bcd92780-f76f-11ea-9be6-6ff51a86460b.png)

3. The Invoice's block name is not displayed
**Before**
![image](https://user-images.githubusercontent.com/16067358/93223339-3f61e700-f770-11ea-9b51-1907423dc08c.png)
**After**
![image](https://user-images.githubusercontent.com/16067358/93223378-4b4da900-f770-11ea-9dd7-d468f0b9f4d1.png)

Thanks to check and feedback.

ks129 committed:
Add invoice download icon as different column in outstanding grid

ks129 commented:
@khouloudbelguith I can't find how to align column to center in new system. In mobile, you have to use Horizontal Scroll. Added invoice as different block. I don't understand issue 4, because it's not possible to change risk as this is connected with customer, but this is displaying invoices.

ks129 commented:
Checks is now fixed

Progi1984 reviewed (commented):
A last comment and it's going to QA

ks129 commented:
@Progi1984 What comment you mean?

PululuK commented:
> @Progi1984 What comment you mean?

Here https://github.com/PrestaShop/PrestaShop/pull/20288#discussion_r503833696

ks129 committed:
Move risks getting to constructor for outstanding grid definition

ks129 commented:
@Progi1984 Comment applied

ks129 committed:
Remove unnecessary trailing spaces from outstanding grid definition

Progi1984 commented on src/PrestaShopBundle/Resources/config/services/core/grid/grid_definition_factory.yml:
```diff
@@ -271,3 +271,10 @@ services:
+        class: 'PrestaShop\PrestaShop\Core\Grid\Definition\Factory\OutstandingGridDefinitionFactory'
+        parent: 'prestashop.core.grid.definition.factory.abstract_grid_definition'
+        arguments:
+            - '@prestashop.adapter.legacy.configuration'
```
```suggestion
            - '@prestashop.adapter.legacy.configuration'
            - "@=service('prestashop.adapter.legacy.context').getContext().language.id"
```

Progi1984 commented on src/Core/Grid/Definition/Factory/OutstandingGridDefinitionFactory.php:
```diff
@@ -0,0 +1,282 @@
+     */
+    private $risks;
+
+    /**
+     * @param HookDispatcherInterface $hookDispatcher
+     * @param ConfigurationInterface $configuration
+     */
+    public function __construct(HookDispatcherInterface $hookDispatcher, ConfigurationInterface $configuration)
+    {
+        parent::__construct($hookDispatcher);
+
+        $this->configuration = $configuration;
+        foreach (Risk::getRisks(Context::getContext()->language->id) as $risk) {
+            /* @var $risk Risk */
+            $this->risks[$risk->name] = $risk->id;
+        }
+    }
```
```suggestion
    /**
     * @param HookDispatcherInterface $hookDispatcher
     * @param ConfigurationInterface $configuration
     * @param int $languageId
     */
    public function __construct(HookDispatcherInterface $hookDispatcher, ConfigurationInterface $configuration, int $languageId)
    {
        parent::__construct($hookDispatcher);

        $this->configuration = $configuration;
        foreach (Risk::getRisks($languageId) as $risk) {
            /* @var $risk Risk */
            $this->risks[$risk->name] = $risk->id;
        }
    }
```

ks129 committed:
Provide context language ID from service for outstanding grid

Co-authored-by: Progi1984 <progi1984@gmail.com>

ks129 committed:
Apply providing language ID from service to outstanding grid

Co-authored-by: Progi1984 <progi1984@gmail.com>

atomiix commented on src/Core/Grid/Definition/Factory/OutstandingGridDefinitionFactory.php:
```diff
@@ -0,0 +1,283 @@
+
+namespace PrestaShop\PrestaShop\Core\Grid\Definition\Factory;
+
+use Context;
```
```suggestion
```

atomiix committed:
Php-cs-fixer

khouloudbelguith commented:
Hi @ks129,

I check the PR but, all [old bugs](https://github.com/PrestaShop/PrestaShop/pull/20288#issuecomment-686560282) are displayed again.
I checked commits => ok
![image](https://user-images.githubusercontent.com/16067358/98124300-4d500080-1eb3-11eb-9c8a-8fd31f0e94c5.png)

1. In legacy, there is no sort action and filter box for "Current outstanding",
![image](https://user-images.githubusercontent.com/16067358/98123556-4d033580-1eb2-11eb-923b-0c82c393172f.png)
2. In legacy, there is no sort action for the "Risk"
![image](https://user-images.githubusercontent.com/16067358/98123725-85a30f00-1eb2-11eb-98a7-bf29c5bee608.png)
3. The sort buttons do not work for all field
![image](https://user-images.githubusercontent.com/16067358/98123924-d286e580-1eb2-11eb-87e4-e5fa92c859b6.png)
4. the Reset button doesn't work
![image](https://user-images.githubusercontent.com/16067358/98124088-0104c080-1eb3-11eb-9bd2-f470e6666872.png)
5. If you click View invoice => a delivery is downloaded and not an invoice
6. If you reload the page, we have an error in the console
![image](https://user-images.githubusercontent.com/16067358/98124189-2396d980-1eb3-11eb-8ec4-52fbeff317d5.png)

Thanks!

ks129 committed:
Add outstanding JS files to webpack

ks129 committed:
Remove unnecessary risk sort and current outstanding sort, filter

ks129 committed:
Add default value for outstanding grid risk column

ks129 committed:
Fix outstanding grid invoice field route

ks129 committed:
Add support for multistore for outstanding grid

ks129 commented:
@khouloudbelguith Sorry about delay, but now all these things (except console error, that I don't see at all) should be fixed.

Progi1984 commented on src/Core/Grid/Definition/Factory/OutstandingGridDefinitionFactory.php:
```diff
@@ -0,0 +1,278 @@
+    private $configuration;
+
+    /**
+     * @var array
```
```suggestion
     * @var array<string, id>
```

Progi1984 commented on src/Core/Grid/Query/OutstandingQueryBuilder.php:
```diff
@@ -0,0 +1,218 @@
+     * @param QueryBuilder $qb
+     * @param SearchCriteriaInterface $criteria
+     */
+    private function applySorting(QueryBuilder $qb, SearchCriteriaInterface $criteria)
```
```suggestion
    private function applySorting(QueryBuilder $qb, SearchCriteriaInterface $criteria): void
```

Progi1984 commented on src/Core/Grid/Query/OutstandingQueryBuilder.php:
```diff
@@ -0,0 +1,218 @@
+     * @param QueryBuilder $qb
+     * @param array $filters
+     */
+    private function applyFilters(QueryBuilder $qb, array $filters)
```
```suggestion
    private function applyFilters(QueryBuilder $qb, array $filters): void
```

Progi1984 commented on src/Core/Grid/Query/OutstandingQueryBuilder.php:
```diff
@@ -0,0 +1,218 @@
+     *
+     * @return QueryBuilder
+     */
+    private function getBaseQueryBuilder(SearchCriteriaInterface $criteria)
```
```suggestion
    private function getBaseQueryBuilder(SearchCriteriaInterface $criteria): QueryBuilder
```

Progi1984 reviewed (commented):
Some feedbacks

ks129 committed:
Add QueryBuilder return type to getBaseQueryBuilder of OutstandingQueryBuilder

Co-authored-by: Progi1984 <progi1984@gmail.com>

ks129 committed:
Add void return type for applyFilters

Co-authored-by: Progi1984 <progi1984@gmail.com>

ks129 committed:
Add void return type for applySorting

Co-authored-by: Progi1984 <progi1984@gmail.com>

ks129 committed:
Define array types in PhpDoc

Co-authored-by: Progi1984 <progi1984@gmail.com>

matks commented on src/Core/Grid/Query/OutstandingQueryBuilder.php:
```diff
@@ -0,0 +1,218 @@
+
+        foreach ($filters as $filterName => $filterValue) {
+            if (isset($strictComparisonFilters[$filterName])) {
+                $alias = $strictComparisonFilters[$filterName];
```
👍

ks129 replied:
What do you mean?

matks replied:
I like how you dit it :) so it's just a "well done !"

Using `$strictComparisonFilters` and `$likeComparisonFilters` is a good way to organize this code.

Code review is also about saying when we think something is well crafted 😄

matks commented on src/PrestaShopBundle/Resources/config/routing/admin/sell/customer/outstanding.yml:
```diff
@@ -0,0 +1,16 @@
+  path: /
+  methods: [POST]
+  defaults:
+    _controller: PrestaShopBundle:Admin\Common:searchGrid
```
If you use common search action in this route, then `PrestaShopBundle:Admin/Sell/Customer/Outstanding:search` is not necessary? 🤔

ks129 replied:
Changed this to use custom one, because common one need grid definition extending with special class.

matks replied:
All right 👍

matks reviewed (commented):
Looks good to me except one thing. Can you check my comment above 😄  I'm wondering about searchAction?

ks129 committed:
Use custom search action instead of common one

ks129 commented:
@Progi1984 One of change you suggested break PhpStan.

matks reviewed (dismissed):
I approve the PR 🎉  great work !

ks129 committed:
Fix risks property PhpDoc type

Progi1984 commented on src/Core/Grid/Definition/Factory/OutstandingGridDefinitionFactory.php:
```diff
@@ -0,0 +1,278 @@
+                    ])
+            );
+
+        if ($this->configuration->get('PS_INVOICE')) {
```
```suggestion
        if ($this->configuration->getBoolean('PS_INVOICE')) {
```

Progi1984 commented on src/Core/Grid/Definition/Factory/OutstandingGridDefinitionFactory.php:
```diff
@@ -0,0 +1,278 @@
+            );
+    }
+
+    private function getInvoiceRowAction()
```
```suggestion
    private function getInvoiceRowAction(): RowActionCollection
```

Progi1984 commented on src/Core/Grid/Definition/Factory/OutstandingGridDefinitionFactory.php:
```diff
@@ -0,0 +1,278 @@
+            );
+    }
+
+    private function getViewRowAction()
```
```suggestion
    private function getViewRowAction(): RowActionCollection
```

Progi1984 reviewed (commented):
Some feedbacks

ks129 committed:
Use getBoolean instead get to get always bool

Co-authored-by: Progi1984 <progi1984@gmail.com>

ks129 committed:
Add missing return type

ks129 commented:
@Progi1984 PHPStan is saying that getBoolean doesn't exist.

ks129 committed:
Revert get -> getBoolean change

ks129 committed:
Add missing newlines

atomiix commented on admin-dev/themes/new-theme/.webpack/common.js:
```diff
@@ -89,6 +90,7 @@ module.exports = {
     order_states: './js/pages/order-states',
     order_return_states_form: './js/pages/order-return-states/form',
     order_view: './js/pages/order/view.js',
+    outstanding: './js/pages/outstanding',
```
Entry point `'./js/pages/outstanding'` is used twice, one for `customer_outstanding` and one for `outstanding`. Only one should be used

ks129 committed:
Remove duplicate outstanding JS entry

ks129 committed:
Use customer_outstanding instead outstanding for JS entry

ks129 committed:
Fix JS file name of outstanding JS in Twig template

sowbiba commented on src/PrestaShopBundle/Resources/views/Admin/Common/Grid/Columns/Content/badge.html.twig:
```diff
@@ -24,12 +24,18 @@
     {{ record[column.options.field] }}
   </span>
+{% elseif record[column.options.field] is not empty and column.options.color_field in record|keys and record[column.options.color_field] is not empty %}
+  {% set textColor = is_color_bright(record[column.options.color_field]) ? '#383838' : 'white' %}
```
You could use classes here 
{% set textClass = is_color_bright(record[column.options.color_field]) ? 'text-dark' : 'text-bright' %}
and <span class="badge rounded {{ textClass }}" ...>

But as we already have inline style for the background-color, no need :)

khouloudbelguith commented:
Hi @ks129,

I have some issues:

0. You have already more than one invoice for John Doe previously
1. Go to BO > Shop Parameters > Orders Settings > Enable B2B
2. Go to BO > Orders > Invoices > Disable invoices
3. Go to BO > Orders > Order list page > change the status of any order (John Doe) to "Payment accepted"
4. Go to BO > Orders > Invoices > Enable invoices
5. Go to BO > Orders > Order list page > View the previous order > Generate the invoice from the Documents Tab
6. Go to BO > Customers > Outstanding > Download the invoice > see error => `invoices.pdf` is downloaded instead of `IN00002.pdf`

I attached a screen record
https://drive.google.com/file/d/1JMmPvmOHwecUohG7i4acjxwDwwR1R8FP/view?usp=sharing

**Without PR** => ok
https://drive.google.com/file/d/1hh86AvjG92VkeMFVoNRQYg-bxqboauBo/view?usp=sharing


I found this issue (With or Without PR)

1. Go to BO > Shop Parameters > Orders Settings > Enable B2B
2. Check the menu => the Outstanding is not added to the menu => we should reload the page
![image](https://user-images.githubusercontent.com/16067358/107640385-7e8a5a80-6c72-11eb-8099-6dd6b57254ca.png)


In the PS177, It is ok.
![image](https://user-images.githubusercontent.com/16067358/107640302-5c90d800-6c72-11eb-80d5-7f6fc5155fe1.png)


Thanks!

ks129 committed:
Add new PDF generator for generating PDF for single invoice

ks129 committed:
Add route to invoices controller that generates single invoice PDF

ks129 committed:
Change outstanding grid PDF generation route

ks129 commented:
@khouloudbelguith First issue is fixed by making a new route (and generator) to invoices controller that generates PDF of **single** invoice. The route that was used before generated based on order ID, but that meant if orders have more than one invoice, then **all** of them would be downloaded. Now with the new route, this doesn't happen anymore. I put this to the invoices controller because I think this is the best place for it (doesn't have much sense to have it in the outstanding controller).

About the second issue, I'm unable to detect what's wrong and this is also a little big out-of-scope for this PR.

PierreRambaud commented on src/PrestaShopBundle/Controller/Admin/Sell/Order/InvoicesController.php:
```diff
@@ -144,4 +144,20 @@ private function processForm(FormHandlerInterface $formHandler, Request $request
+
+        // When using legacy generator,
+        // we want to be sure that displaying PDF is the last thing this controller will do
+        die();
```
I'm afraid of side effect. Could you use BinaryResponse? or if needed can be catch by an ob_start() :thinking:

ks129 replied:
https://github.com/PrestaShop/PrestaShop/blob/develop/src/PrestaShopBundle/Controller/Admin/Sell/Order/OrderController.php#L284-L291 I made this same way like orders controller do.

PierreRambaud replied:
What the fuuu*** :scream:

khouloudbelguith commented:
Hi @ks129,

It is ok :heavy_check_mark: 

- Help button => ok
- Sort by all fields ASC & DSC => OK
- Filter by all fields one by one & all => ok / Filter no records => ok
- Pagination => ok
- View invoice => OK
- View Order => ok
- Multistore behavior => OK
- Mobile => OK

https://drive.google.com/file/d/1HI6aSQMiDrinb2OetL0yJTO9h_jp6LWw/view?usp=sharing

Thank you!

Progi1984 commented:
Thanks @ks129 & @khouloudbelguith
